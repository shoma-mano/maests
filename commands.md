# Measts Commands

# Measts Commands

## Table of Contents

- [Initialize Flow](#initialize-flow)
- [Taps and Presses](#taps-and-presses)
- [Swipes and Scrolls](#swipes-and-scrolls)
- [Text Input](#text-input)
- [Navigation and Links](#navigation-and-links)
- [Assertions](#assertions)
- [Waiting](#waiting)
- [Keyboard Actions](#keyboard-actions)
- [Device Volume Controls](#device-volume-controls)
- [Miscellaneous Actions](#miscellaneous-actions)
- [Repeat Actions](#repeat-actions)

### Initialize Flow

- **Launch the Application**
```typescript
    M.initFlow({ appId: "com.example.app" })
```

- **Clear Application State**
```typescript
    M.clearState({ appId: "com.example.app" })
```

- **Clear Keychain**
```typescript
    M.clearKeychain()
```

### Taps and Presses**

Taps and Presses have optional properties available for configuring various Maestro commands, designed to enhance the flexibility and customization of user interactions. The properties are described below.

#### Properties

### TapProps

The `TapProps` interface defines optional properties for customizing tap actions in Maestro commands, such as `tapOn`, `tapOnText`, and `tapOnPoint`.

| Property                | Type     | Default | Description                                                                                                  |
|-------------------------|----------|---------|--------------------------------------------------------------------------------------------------------------|
| `retryTapIfNoChange`    | boolean  | `true`  | If set to `true`, the tap action will be retried if no change is detected on the first attempt.              |
| `repeat`                | number   | -       | Specifies the number of times to repeat the tap action.                                                      |
| `waitToSettleTimeoutMs` | number   | -       | Time in milliseconds to wait for the element to settle after tapping, useful for handling UI transition time. |

#### Example Usage:
```typescript
M.tapOn("submitButton", { retryTapIfNoChange: false, repeat: 2, waitToSettleTimeoutMs: 500 });

---

### WaitProps

```markdown
### WaitProps

The `WaitProps` interface provides optional properties for configuring wait behavior in specific commands, such as `waitForAndTapOn`, `waitUntilVisible`, and `waitUntilNotVisible`.

| Property   | Type    | Default | Description                                                                                     |
|------------|---------|---------|-------------------------------------------------------------------------------------------------|
| `maxWait`  | number  | 5000    | The maximum time (in milliseconds) to wait for an element or condition before continuing.       |

#### Example Usage:
```typescript
M.waitForAndTapOn("startButton", { maxWait: 3000 });


---

### PointProps

```markdown
### PointProps

The `PointProps` interface specifies coordinates for point-based tap or long-press actions, enabling precise control of on-screen actions.

| Property | Type                | Description                                                                      |
|----------|----------------------|----------------------------------------------------------------------------------|
| `x`      | `number | string`   | The x-coordinate for the point to tap on. Can be an absolute number or percentage.|
| `y`      | `number | string`   | The y-coordinate for the point to tap on. Can be an absolute number or percentage.|

#### Example Usage:
```typescript
M.tapOnPoint({ x: "50%", y: "50%" }, { waitToSettleTimeoutMs: 200 });


- **Tap on Element by ID**

```typescript
    M.tapOn("submitButton", { retryTapIfNoChange: true, repeat: 2, waitToSettleTimeoutMs: 500 })
```
- **Tap on Visible Text**
```typescript
    M.tapOnText("Submit", { retryTapIfNoChange: true })
```
- **Tap on Point**
```typescript
    M.tapOnPoint({ x: "50%", y: "50%" }, { waitToSettleTimeoutMs: 200 })
```
- **Wait for Element to Appear and Tap**
```typescript
    M.waitForAndTapOn("submitButton", 3000)
```
- **Long Press on Element by ID**
```typescript
    M.longPressOn("submitButton")
```
- **Long Press on Text**
```typescript
    M.longPressOnText("Submit")
```

### Swipes and Scrolls

- **Swipe in a Direction**
```typescript
    M.swipeLeft()
    M.swipeRight()
    M.swipeUp()
    M.swipeDown()
```

- **Swipe from Point to Point**
```typescript
   M.swipe({ x: "10%", y: "20%" }, { x: "90%", y: "80%" })
```

- **Scroll Screen**
```typescript
    M.scroll()
```

- **Scroll Until Element is Visible**
```typescript
   M.scrollUntilVisible("scrollTarget")
```

### Text Input

- **Input Text into an Element**
```typescript
   M.inputText("Hello, World!", "textFieldId")
```

- **Input Random Name**
```typescript
    M.inputRandomName("nameFieldId")
```

- **Input Random Number**
```typescript
    M.inputRandomNumber("numberFieldId")
```

- **Input Random Email**
```typescript
    M.inputRandomEmail("emailFieldId")
```

- **Input Random Text**
```typescript
    M.inputRandomText("textFieldId")
```

- **Input Random Text**
```typescript
    M.inputRandomText("textFieldId")
```

- **Erase Text**
```typescript
    M.eraseText(10, "textFieldId")
```

### Navigation and Links

- **Open a Link**
```typescript
    M.openLink("https://example.com")
```

- **Navigate to Path Using Deep Link Base**
```typescript
   M.navigate("/home")
```

### Assertions

- **Assert Element is Visible**
```typescript
  M.assertVisible("submitButton", true)
```

- **Assert Element is Not Visible**
```typescript
  M.assertNotVisible("submitButton")
```

### Waiting
- **Wait for Animation End**
```typescript
  M.waitForAnimationEnd(3000)
```

- **Wait Until Element is Visible**
```typescript
  M.waitUntilVisible("submitButton", 5000)
```

- **Wait Until Element is Not Visiblee**
```typescript
  M.waitUntilNotVisible("loadingSpinner", 5000)
```

- **Wait a Specified Number of Milliseconds**
```typescript
  M.wait(1000)
```

### Keyboard Actions

- **Hide Keyboard**
```typescript
  M.hideKeyboard()
```

- **Press Enter**
```typescript
  M.pressEnter()
```

- **Press Home Button**
```typescript
  M.pressHomeButton()
```

- **Press Lock Button**
```typescript
  M.pressLockButton()
```

- **Press Back Button**
```typescript
  M.back()
```

### Device Volume Controls

- **Increase Volume**
```typescript
    M.volumeUp()
```

- **Decrease Volume**
```typescript
   M.volumeDown()
```

### Miscellaneous Actions

- **Take Screenshot**
```typescript
    M.screenshot("my-screenshot.png")
```

- **Copy Text from an Element**
```typescript
    M.copyTextFrom("textElementId")
```

- **Assert Condition is True**
```typescript
    M.assertTrue("2 + 2 === 4")
```

### Repeat Actions

- **Repeat Actions a Specified Number of Times**
```typescript
    M.repeat(3, () => {
    M.tapOn("submitButton");
    M.wait(500)
});
```

- **Repeat Actions While Element is Visibles**
```typescript
    M.repeatWhileVisible("loadingSpinner", () => {
    M.wait(1000)
});
```
- **Repeat Actions While Element is Not Visible**
```typescript
    M.repeatWhileNotVisible("submitButton", () => {
    M.wait(1000);
});
```






































