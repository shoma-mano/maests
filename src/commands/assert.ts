import { stringify } from "yaml";
import { addOut, getOut } from "../out";

type AssertProps = {
  text?: string;
  id?: string;
  enabled?: boolean;
  checked?: boolean;
  focused?: boolean;
  selected?: boolean;
};

export const assertVisible = (props: AssertProps) => {
  const commands = [{ assertVisible: { ...props } }];
  addOut(stringify(commands));
};

if (import.meta.vitest) {
  it("assertVisible", () => {
    assertVisible({ id: "com.android.systemui:id/battery" });
    expect(getOut()).toMatchInlineSnapshot(`
      "- assertVisible:
          id: com.android.systemui:id/battery
      "
    `);
  });
}

export const assertNotVisible = (props: AssertProps) => {
  const commands = [{ assertNotVisible: { ...props } }];
  addOut(stringify(commands));
};

if (import.meta.vitest) {
  it("assertNotVisible", () => {
    assertNotVisible({ id: "com.android.systemui:id/battery" });
    expect(getOut()).toMatchInlineSnapshot(`
      "- assertNotVisible:
          id: com.android.systemui:id/battery
      "
    `);
  });
}
