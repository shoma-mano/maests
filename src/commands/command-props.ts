export interface WaitProps {
  maxWait?: number;
}

export interface PointProps {
  x: number | string;
  y: number | string;
}
