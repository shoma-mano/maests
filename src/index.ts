#!/usr/bin/env node
import fs, { writeFileSync } from "fs";
import path, { dirname, join } from "path";
import createJiti from "jiti";
import dotenv from "dotenv";
import { consola } from "consola";
import { rewriteCode } from "./rewriteCode";
import { defineCommand, runMain } from "citty";
import { createYamlOutPath } from "./utils";
import { execSync } from "child_process";
import { getTsconfig } from "get-tsconfig";

const main = defineCommand({
  args: {
    path: {
      type: "positional",
      required: true,
    },
    device: {
      type: "string",
      alias: "d",
      description: "Device to run the test on",
    },
  },
  async run({ args }) {
    loadEnv();
    const cwd = process.cwd();

    // create temp file
    let flowPath = args.path;
    if (flowPath.startsWith(cwd)) {
      flowPath = flowPath.replace(`${cwd}/`, "");
    }
    const yamlOutPath = createYamlOutPath(flowPath);
    let code = fs.readFileSync(flowPath, "utf-8");
    code = rewriteCode({ code, yamlOutPath });
    const tempFilePath = join(cwd, `${flowPath.replace(".ts", ".temp.ts")}`);
    writeFileSync(tempFilePath, code);

    // create jiti instance
    const { config, path } = getTsconfig();
    const normalizedAlias = Object.fromEntries(
      Object.entries(config?.compilerOptions?.paths || {}).map(
        ([key, value]) => {
          const normalizedKey = key.replace("/*", "");
          const normalizedValue = value[0].replace("/*", "");
          return [normalizedKey, join(dirname(path), normalizedValue)];
        }
      )
    );
    const jiti = createJiti(cwd, {
      esmResolve: true,
      experimentalBun: true,
      alias: normalizedAlias,
    });

    // execute temp file
    try {
      await jiti(tempFilePath);
      consola.success(`Created Yaml to ${yamlOutPath} âœ”`);
    } catch (e) {
      console.error(e);
      fs.unlinkSync(tempFilePath);
      process.exit(1);
    }
    fs.unlinkSync(tempFilePath);

    // run maestro test
    const command = `maestro ${
      args.device ? `--device ${args.device}` : ""
    } test  ${yamlOutPath}`;
    try {
      execSync(command, {
        stdio: "inherit",
        env: process.env,
      });
    } catch (e) {
      console.error(e);
      process.exit(1);
    }
  },
});
if (!import.meta.vitest) runMain(main);

function loadEnv() {
  let currentPath = process.cwd();
  let dotEnvPath = "";
  while (currentPath !== "/") {
    const files = fs.readdirSync(currentPath);
    if (files.includes(".env")) {
      dotEnvPath = path.join(currentPath, ".env");
      break;
    }
    currentPath = path.join(currentPath, "../");
  }
  if (dotEnvPath) {
    consola.info(`Found .env file at ${dotEnvPath}`);
    dotenv.config({ path: dotEnvPath });
  }
}
