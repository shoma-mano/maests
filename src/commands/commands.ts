import { tapOn, tapOnPoint, tapOnText, waitForAndTapOn } from "./tap";
import { runFlow, runScript } from "./run";
import { clearState, initFlow, launchApp } from "./init";
import { repeat, repeatWhileVisible, repeatWhileNotVisible } from "./repeat";
import { assertNotVisible, assertVisible } from "./assert";
import { addOut } from "../out";
import { PointProps } from "../type";

// Utility function for indenting except last line break
export const indentExceptLastLineBreak = (str: string) => {
  return str.replace(/\n(?=.*[\n])/g, "\n        ");
};

export const space = "    ";

// Main translator functions
export const MaestroTranslators = {
  /**
   * Initializes the test flow with optional configuration.
   * @param config - Optional configuration with appId and other environment variables.
   */
  initFlow,
  /**
   * Launches the application with optional configuration settings.
   * @param config - Configuration options for appId, state clearing, keychain clearing, and app stopping.
   */
  launchApp,
  /**
   * Clears the state of the current app or the specified app by appId.
   * @param appId - Optional appId to clear state for a specific app.
   */
  clearState,

  runScript,

  runFlow,

  tapOn,

  tapOnText,

  tapOnPoint,

  waitForAndTapOn,

  clearKeychain: () => {
    addOut("- clearKeychain\n");
  },

  longPressOn: (id: string) => {
    addOut(`- longPressOn:\n    id: "${id}"\n`);
  },

  longPressOnPoint: (pointProps: PointProps) => {
    addOut(`- longPressOn:\n    point: ${pointProps.x},${pointProps.y}\n`);
  },

  longPressOnText: (text: string) => {
    addOut(`- longPressOn: ${text}\n`);
  },

  swipeLeft: () => addOut("- swipe:\n    direction: LEFT\n    duration: 400\n"),
  swipeRight: () =>
    addOut("- swipe:\n    direction: RIGHT\n    duration: 400\n"),
  swipeDown: () => addOut("- swipe:\n    direction: DOWN\n    duration: 400\n"),
  swipeUp: () => addOut("- swipe:\n    direction: UP\n    duration: 400\n"),

  swipe: (start: PointProps, end: PointProps) => {
    addOut(
      `- swipe:\n    start: ${start.x}, ${start.y}\n    end: ${end.x}, ${end.y}\n`
    );
  },

  inputText: (text: string, id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputText: ${text}\n`
        : `- inputText: ${text}\n`
    );
  },

  inputRandomName: (id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputRandomPersonName\n`
        : `- inputRandomPersonName\n`
    );
  },

  inputRandomNumber: (id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputRandomNumber\n`
        : `- inputRandomNumber\n`
    );
  },

  copyTextFrom: (id: string) => {
    addOut(`- copyTextFrom:\n    id: "${id}"\n`);
  },

  inputRandomEmail: (id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputRandomEmail\n`
        : `- inputRandomEmail\n`
    );
  },

  inputRandomText: (id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputRandomText\n`
        : `- inputRandomText\n`
    );
  },

  eraseText: (chars: number, id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- eraseText: ${chars ?? 50}\n`
        : `- eraseText: ${chars ?? 50}\n`
    );
  },

  openLink: (url: string) => {
    addOut(`- openLink: ${url}\n`);
  },

  navigate: (path: string) => {
    addOut(`- openLink: ${process.env["deepLinkBase"]}${path}\n`);
  },

  assertVisible,

  assertNotVisible,

  scroll: () => {
    addOut("- scroll\n");
  },

  scrollUntilVisible: (id: string) => {
    addOut(`- scrollUntilVisible:\n    element:\n      id: "${id}"\n`);
  },

  waitForAnimationEnd: (maxWait: number = 5000) => {
    addOut(
      maxWait
        ? `- waitForAnimationToEnd:\n    timeout: ${maxWait}\n`
        : "- waitForAnimationToEnd\n"
    );
  },

  waitUntilVisible: (id: string, maxWait: number) => {
    addOut(
      `- extendedWaitUntil:\n    visible:\n        id: "${id}"\n    timeout: ${
        maxWait ?? 5000
      }\n`
    );
  },

  waitUntilNotVisible: (id: string, maxWait: number) => {
    addOut(
      `- extendedWaitUntil:\n    notVisible:\n        id: "${id}"\n    timeout: ${
        maxWait ?? 5000
      }\n`
    );
  },

  wait: (ms: number) => {
    addOut(
      `- swipe:\n    start: -1, -1\n    end: -1, -100\n    duration: ${ms}\n`
    );
  },

  hideKeyboard: () => {
    addOut("- hideKeyboard\n");
  },

  screenshot: (fileName: string) => {
    addOut(`- takeScreenshot: ${fileName}\n`);
  },

  pressEnter: () => {
    addOut("- pressKey: Enter\n");
  },

  pressHomeButton: () => {
    addOut("- pressKey: Home\n");
  },

  pressLockButton: () => {
    addOut("- pressKey: Lock\n");
  },

  back: () => {
    addOut("- pressKey: back\n");
  },

  volumeDown: () => {
    addOut("- pressKey: volume down\n");
  },

  volumeUp: () => {
    addOut("- pressKey: volume up\n");
  },

  stopApp: ({ appId }: { appId?: string } = {}) => {
    addOut(appId ? `- stopApp: ${appId}\n` : "- stopApp\n");
  },

  repeat,

  repeatWhileVisible,

  repeatWhileNotVisible,

  yaml: (yaml: string) => `${yaml}\n`,

  assertTrue: (condition: string) => {
    addOut(`- assertTrue: ${condition}\n`);
  },
};

export { MaestroTranslators as M };
export { writeYaml } from "../write-yaml";

// utils for user
export const getOutput = (key: string) => "${output." + key + "}";

// runScript Types
declare global {
  namespace http {
    const get: (...args: any) => { body: string };
  }

  const json: <T extends any>(str: string) => T;
  const output: Record<string, string>;
}
