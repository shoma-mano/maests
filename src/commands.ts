import { TapProps, PointProps, WaitProps } from "./command-props";

// Nested command handling
let isNested = false;
let nestedCommands = "";
const handleNest = (func: () => any) => {
  isNested = true;
  func();
  isNested = false;
  const out = nestedCommands;
  nestedCommands = "";
  return out;
};

export let out = "";
export const resetOut = () => {
  out = "";
};
const addOut = (command: string) => {
  if (isNested) nestedCommands += command;
  else out += command;
};

// Helper function to format optional tap properties
/**
 * Formats optional tap properties into the command output format.
 * @param retryTapIfNoChange - If true, retries the tap if no change occurs after the first attempt.
 * @param repeat - The number of times to repeat the tap action.
 * @param waitToSettleTimeoutMs - The time to wait (in milliseconds) for the element to settle after tapping.
 * @returns The formatted string of optional tap properties.
 */
const formatTapProps = ({
  retryTapIfNoChange = true,
  repeat,
  waitToSettleTimeoutMs,
}: TapProps): string => {
  let propsCommand = "";

  if (retryTapIfNoChange === false) propsCommand += `    retryTapIfNoChange: ${retryTapIfNoChange}\n`;
  if (typeof repeat === "number") propsCommand += `    repeat: ${repeat}\n`;
  if (typeof waitToSettleTimeoutMs === "number") propsCommand += `    waitToSettleTimeoutMs: ${waitToSettleTimeoutMs}\n`;

  return propsCommand;
};

type WaitAndTapProps = TapProps & WaitProps;

// Main translator functions
const envAppId = process.env["appId"];
export const MaestroTranslators = {
  /**
   * Initializes the test flow with an optional application ID.
   * @param appId - Optional application ID to override the default environment appId.
   */
  initFlow: ({ appId }: { appId?: string } = {}) => {
    addOut(`appId: ${appId ?? envAppId}\n---\n`);
  },

  /**
   * Launches the application with optional configurations.
   * @param appId - Optional app ID to use for launching the app.
   */
  launchApp: ({ appId }: { appId?: string } = {}) => {
    addOut(`- launchApp:\n    appId: "${appId ?? envAppId}"\n`);
  },

  /**
   * Clears the state of the application.
   * @param appId - Optional app ID to clear a specific application's state.
   */
  clearState: ({ appId }: { appId?: string } = {}) => {
    addOut(appId ? `- clearState: ${appId}\n` : "- clearState\n");
  },

  /**
   * Clears the entire keychain.
   */
  clearKeychain: () => {
    addOut("- clearKeychain\n");
  },

  /**
   * Taps on an element by its test ID with optional retry and repeat configurations.
   * @param id - The test ID of the target element.
   * @param props - Optional properties to customize the tap action.
   */
  tapOn: (id: string, props: TapProps = {}) => {
    let command = `- tapOn:\n    id: "${id}"\n`;
    command += formatTapProps(props);
    addOut(command);
  },

  /**
   * Taps on visible text on the screen with optional retry and repeat configurations.
   * @param text - The visible text to tap on.
   * @param props - Optional properties to customize the tap action.
   */
  tapOnText: (text: string, props: TapProps = {}) => {
    let command = `- tapOn:\n    text: "${text}"\n`;
    command += formatTapProps(props);
    addOut(command);
  },

  /**
   * Taps on a specified point on the screen.
   * @param point - The x and y coordinates of the tap.
   * @param props - Optional properties for customizing the tap action.
   */
  tapOnPoint: (point: PointProps, props: TapProps = {}) => {
    const { x, y } = point;
    let command = `- tapOn:\n    point: ${x},${y}\n`;
    command += formatTapProps(props);
    addOut(command);
  },

  /**
   * Waits for an element by testId to appear, then taps on it.
   * @param id - The testId of the element to wait for and tap.
   * @param props - Optional wait and tap properties, including maxWait for wait-based actions.
   */
  waitForAndTapOn: (id: string, props: WaitAndTapProps = {}) => {
    const { maxWait = 5000 } = props; // Default maxWait only for this function

    let command = `- extendedWaitUntil:\n    visible:\n        id: "${id}"\n    timeout: ${maxWait}\n`;
    command += `- tapOn:\n    id: "${id}"\n`;
    command += formatTapProps(props);
    addOut(command);
  },

  /**
   * Long presses on an element by its testId.
   * @param id - The testId of the element to long press on.
   */
  longPressOn: (id: string) => {
    addOut(`- longPressOn:\n    id: "${id}"\n`);
  },

  /**
   * Long presses on a specified point on the screen.
   * @param point - The x and y coordinates for the long press.
   */
  longPressOnPoint: ({ x, y }: PointProps) => {
    addOut(`- longPressOn:\n    point: ${x},${y}\n`);
  },

  /**
   * Long presses on a specified text on the screen.
   * @param text - The text to long press.
   */
  longPressOnText: (text: string) => {
    addOut(`- longPressOn: ${text}\n`);
  },

  /**
   * Swipes in the specified direction from the center of the screen.
   */
  swipeLeft: () => addOut("- swipe:\n    direction: LEFT\n    duration: 400\n"),
  swipeRight: () => addOut("- swipe:\n    direction: RIGHT\n    duration: 400\n"),
  swipeDown: () => addOut("- swipe:\n    direction: DOWN\n    duration: 400\n"),
  swipeUp: () => addOut("- swipe:\n    direction: UP\n    duration: 400\n"),

  /**
   * Swipes from a starting to an ending point.
   * @param start - Starting coordinates of the swipe.
   * @param end - Ending coordinates of the swipe.
   */
  swipe: (start: PointProps, end: PointProps) => {
    addOut(`- swipe:\n    start: ${start.x}, ${start.y}\n    end: ${end.x}, ${end.y}\n`);
  },

  /**
   * Inputs text into a focused element or the specified input by testId.
   * @param text - The text to input.
   * @param id - Optional testId of the target input element.
   */
  inputText: (text: string, id?: string) => {
    addOut(id ? `- tapOn:\n    id: "${id}"\n- inputText: ${text}\n` : `- inputText: ${text}\n`);
  },

  /**
   * Inputs a random name into a focused input or specified by testId.
   * @param id - Optional testId of the target input element.
   */
  inputRandomName: (id?: string) => {
    addOut(id ? `- tapOn:\n    id: "${id}"\n- inputRandomPersonName\n` : `- inputRandomPersonName\n`);
  },

  /**
   * Inputs a random number into a focused input or specified by testId.
   * @param id - Optional testId of the target input element.
   */
  inputRandomNumber: (id?: string) => {
    addOut(id ? `- tapOn:\n    id: "${id}"\n- inputRandomNumber\n` : `- inputRandomNumber\n`);
  },

  /**
   * Copies text from an element by its testId.
   * @param id - The testId of the element.
   */
  copyTextFrom: (id: string) => {
    addOut(`- copyTextFrom:\n    id: "${id}"\n`);
  },

  /**
   * Inputs a random email into a focused input or specified by testId.
   * @param id - Optional testId of the target input element.
   */
  inputRandomEmail: (id?: string) => {
    addOut(id ? `- tapOn:\n    id: "${id}"\n- inputRandomEmail\n` : `- inputRandomEmail\n`);
  },

  /**
   * Inputs random text into a focused input or specified by testId.
   * @param id - Optional testId of the target input element.
   */
  inputRandomText: (id?: string) => {
    addOut(id ? `- tapOn:\n    id: "${id}"\n- inputRandomText\n` : `- inputRandomText\n`);
  },

  /**
   * Erases a specified number of characters from an input.
   * @param chars - Number of characters to erase.
   * @param id - Optional testId of the target input element.
   */
  eraseText: (chars: number, id?: string) => {
    addOut(id ? `- tapOn:\n    id: "${id}"\n- eraseText: ${chars ?? 50}\n` : `- eraseText: ${chars ?? 50}\n`);
  },

  /**
   * Opens a specified URL or deep link.
   * @param url - The URL or deep link to open.
   */
  openLink: (url: string) => {
    addOut(`- openLink: ${url}\n`);
  },

  /**
   * Navigates to a specific path using the deep link base.
   * @param path - The path to navigate.
   */
  navigate: (path: string) => {
    addOut(`- openLink: ${process.env["deepLinkBase"]}${path}\n`);
  },

  /**
   * Asserts that an element by testId is visible.
   * @param id - The testId of the element.
   * @param enabled - Optional; checks if the element is enabled.
   */
  assertVisible: (id: string, enabled: boolean = false) => {
    addOut(enabled ? `- assertVisible:\n    id: "${id}"\n    enabled: true\n` : `- assertVisible:\n    id: "${id}"\n`);
  },

  /**
   * Asserts that an element by testId is not visible.
   * @param id - The testId of the element.
   */
  assertNotVisible: (id: string) => {
    addOut(`- assertNotVisible:\n    id: "${id}"\n`);
  },

  /**
   * Scrolls the screen.
   */
  scroll: () => {
    addOut("- scroll\n");
  },

  /**
   * Scrolls until an element with the given testId is visible.
   * @param id - The testId of the element.
   */
  scrollUntilVisible: (id: string) => {
    addOut(`- scrollUntilVisible:\n    element:\n      id: "${id}"\n`);
  },

  /**
   * Waits until an animation/video finishes and screen becomes static.
   * @param maxWait - Optional: Max timeout after which flow continues. Defaults to 5000ms
   */
  waitForAnimationEnd: (maxWait: number = 5000) => {
    const command = maxWait ? `- waitForAnimationToEnd:\n    timeout: ${maxWait}\n` : "- waitForAnimationToEnd\n";
    addOut(command);
  },

  /**
   * Waits until an element by testId is visible.
   * @param id - The testId of the element.
   * @param maxWait - Max wait time in milliseconds.
   */
  waitUntilVisible: (id: string, maxWait: number) => {
    addOut(`- extendedWaitUntil:\n    visible:\n        id: "${id}"\n    timeout: ${maxWait ?? 5000}\n`);
  },

  /**
   * Waits until an element by testId is not visible.
   * @param id - The testId of the element.
   * @param maxWait - Max wait time in milliseconds.
   */
  waitUntilNotVisible: (id: string, maxWait: number) => {
    addOut(`- extendedWaitUntil:\n    notVisible:\n        id: "${id}"\n    timeout: ${maxWait ?? 5000}\n`);
  },

  /**
   * Waits a specified number of milliseconds.
   * @param ms - Number of milliseconds to wait.
   */
  wait: (ms: number) => {
    addOut(`- swipe:\n    start: -1, -1\n    end: -1, -100\n    duration: ${ms}\n`);
  },

  /**
   * Dismisses the software keyboard.
   */
  hideKeyboard: () => {
    addOut("- hideKeyboard\n");
  },

  /**
   * Takes a screenshot and stores it with the specified filename.
   * @param fileName - The name to save the screenshot under.
   */
  screenshot: (fileName: string) => {
    addOut(`- takeScreenshot: ${fileName}\n`);
  },

  /**
   * Presses the enter key on the software keyboard.
   */
  pressEnter: () => {
    addOut("- pressKey: Enter\n");
  },

  /**
   * Presses the home button on the device.
   */
  pressHomeButton: () => {
    addOut("- pressKey: Home\n");
  },

  /**
   * Presses the lock button on the device.
   */
  pressLockButton: () => {
    addOut("- pressKey: Lock\n");
  },

  /**
   * Presses the Android back button.
   */
  back: () => {
    addOut("- pressKey: back\n");
  },

  /**
   * Decreases the device volume.
   */
  volumeDown: () => {
    addOut("- pressKey: volume down\n");
  },

  /**
   * Increases the device volume.
   */
  volumeUp: () => {
    addOut("- pressKey: volume up\n");
  },

  /**
   * Stops the current app or specified app ID.
   * @param appId - Optional; the app ID to stop.
   */
  stopApp: ({ appId }: { appId?: string } = {}) => {
    appId = appId ?? envAppId;
    addOut(appId ? `- stopApp: ${appId}\n` : "- stopApp\n");
  },

  /**
   * Repeats specified actions a given number of times.
   * @param times - Number of repetitions.
   * @param func - Actions to repeat.
   */
  repeat: (times: number, func: () => void) => {
    const out = handleNest(func);
    addOut(`- repeat:\n    times: ${times}\n    commands:\n        ${out.replace(/\n/g, "\n        ")}`);
  },

  /**
   * Repeats actions while an element by testId is visible.
   * @param id - The testId of the element.
   * @param func - Actions to repeat.
   */
  repeatWhileVisible: (id: string, func: () => void) => {
    const out = handleNest(func);
    addOut(`- repeat:\n    while:\n        visible:\n            id: "${id}"\n    commands:\n        ${out.replace(/\n/g, "\n        ")}`);
  },

  /**
   * Repeats actions while an element by testId is not visible.
   * @param id - The testId of the element.
   * @param func - Actions to repeat.
   */
  repeatWhileNotVisible: (id: string, func: () => void) => {
    const out = handleNest(func);
    addOut(`- repeat:\n    while:\n        notVisible:\n            id: "${id}"\n    commands:\n        ${out.replace(/\n/g, "\n        ")}`);
  },

  /**
   * Inserts inline YAML code for specialized commands.
   * @param yaml - The inline YAML to insert.
   */
  yaml: (yaml: string) => `${yaml}\n`,

  /**
   * Checks if a condition is true.
   * @param condition - The condition to assert.
   */
  assertTrue: (condition: string) => {
    addOut(`- assertTrue: ${condition}\n`);
  },
};

export { MaestroTranslators as M };
export { writeYaml } from "./write-yaml";
