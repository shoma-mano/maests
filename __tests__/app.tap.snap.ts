import { describe, it, expect, beforeEach } from "vitest";
import { MaestroTranslators, resetOut, out } from "../src/commands";

const getOutput = () => {
  const currentOutput = out;
  resetOut(); // Reset output for clean snapshots in each test
  return currentOutput;
};

describe("MaestroTranslators Tap Command Snapshots", () => {
  beforeEach(() => {
    resetOut();
  });

  it("tapOn with id only", () => {
    MaestroTranslators.tapOn("testId");
    expect(getOutput()).toMatchSnapshot();
  });

  it("tapOn with all properties", () => {
    MaestroTranslators.tapOn("testId", { retryTapIfNoChange: false, repeat: 3, waitToSettleTimeoutMs: 200 });
    expect(getOutput()).toMatchSnapshot();
  });

  it("tapOnText with text and properties", () => {
    MaestroTranslators.tapOnText("sample text", { retryTapIfNoChange: true, repeat: 2 });
    expect(getOutput()).toMatchSnapshot();
  });

  it("tapOnPoint with coordinates and properties", () => {
    MaestroTranslators.tapOnPoint({ x: "50%", y: "50%" }, { repeat: 4, waitToSettleTimeoutMs: 500 });
    expect(getOutput()).toMatchSnapshot();
  });

  it("tapOn with all properties", () => {
    MaestroTranslators.tapOn("testId", { retryTapIfNoChange: false});
    expect(getOutput()).toMatchSnapshot();
  });

  it("waitForAndTapOn with id and maxWait", () => {
    MaestroTranslators.waitForAndTapOn("testId", 3000);
    expect(getOutput()).toMatchSnapshot();
  });

  it("waitForAndTapOn with all properties", () => {
    MaestroTranslators.waitForAndTapOn("testId", 3000, { retryTapIfNoChange: false, waitToSettleTimeoutMs: 400 });
    expect(getOutput()).toMatchSnapshot();
  });

  it("tapOn with formatted tap props", () => {
    MaestroTranslators.tapOn("testId", {
      retryTapIfNoChange: false,
      repeat: 3,
      waitToSettleTimeoutMs: 500,
    });
    expect(getOutput()).toMatchSnapshot();
  });

});
