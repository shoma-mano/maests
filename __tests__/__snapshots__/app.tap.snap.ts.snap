// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MaestroTranslators Tap Command Snapshots > tapOn with all properties 1`] = `
"- tapOn:
    id: "testId"
    retryTapIfNoChange: false
    repeat: 3
    waitToSettleTimeoutMs: 200
"
`;

exports[`MaestroTranslators Tap Command Snapshots > tapOn with all properties 2`] = `
"- tapOn:
    id: "testId"
    retryTapIfNoChange: false
"
`;

exports[`MaestroTranslators Tap Command Snapshots > tapOn with formatted tap props 1`] = `
"- tapOn:
    id: "testId"
    retryTapIfNoChange: false
    repeat: 3
    waitToSettleTimeoutMs: 500
"
`;

exports[`MaestroTranslators Tap Command Snapshots > tapOn with id only 1`] = `
"- tapOn:
    id: "testId"
"
`;

exports[`MaestroTranslators Tap Command Snapshots > tapOnPoint with coordinates and properties 1`] = `
"- tapOn:
    point: 50%,50%
    repeat: 4
    waitToSettleTimeoutMs: 500
"
`;

exports[`MaestroTranslators Tap Command Snapshots > tapOnText with text and properties 1`] = `
"- tapOn:
    text: "sample text"
    repeat: 2
"
`;

exports[`MaestroTranslators Tap Command Snapshots > waitForAndTapOn with all properties 1`] = `
"- extendedWaitUntil:
    visible:
        id: "testId"
    timeout: 5000
- tapOn:
    id: "testId"
"
`;

exports[`MaestroTranslators Tap Command Snapshots > waitForAndTapOn with id and maxWait 1`] = `
"- extendedWaitUntil:
    visible:
        id: "testId"
    timeout: 5000
- tapOn:
    id: "testId"
"
`;
